gnovm/pkg/doc/dirs.go

```diff
@@ -0,0 +1,167 @@
+// Mostly copied from go source at tip, cmd/doc/dirs.go.
+// https://github.com/golang/go/blob/d922c0a8f5035b0533eb6e912ffd7b85487e3942/src/cmd/doc/dirs.go
+//
+// Copyright 2015 The Go Authors. All rights reserved.
+
+package doc
+
+import (
+	"log"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// A Dir describes a directory holding code by specifying
+// the expected import path and the file system directory.
+type Dir struct {
+	importPath string // import path for that dir
+	dir        string // file system directory
+}
+
+// Dirs is a structure for scanning the directory tree.
+// Its Next method returns the next Go source directory it finds.
+// Although it can be used to scan the tree multiple times, it
+// only walks the tree once, caching the data it finds.
+type Dirs struct {
+	scan   chan Dir // Directories generated by walk.
+	hist   []Dir    // History of reported Dirs.
+	offset int      // Counter for Next.
+}
+
+// NewDirs begins scanning the given stdlibs directory.
+func NewDirs(dirs ...string) *Dirs {
+	d := &Dirs{
+		// prealloc to 256 -- this will generally be a few hundred dirs
+		// so starting off like this spares us log2(128) extra allocations in the general case.
+		hist: make([]Dir, 0, 256),
+		scan: make(chan Dir),
+	}
+	go d.walk(dirs)
+	return d
+}
+
+// Reset puts the scan back at the beginning.
+func (d *Dirs) Reset() {
+	d.offset = 0
+}
+
+// Next returns the next directory in the scan. The boolean
+// is false when the scan is done.
+func (d *Dirs) Next() (Dir, bool) {
+	if d.offset < len(d.hist) {
+		dir := d.hist[d.offset]
+		d.offset++
+		return dir, true
+	}
+	dir, ok := <-d.scan
+	if !ok {
+		return Dir{}, false
+	}
+	d.hist = append(d.hist, dir)
+	d.offset++
+	return dir, ok
+}
+
+// walk walks the trees in the given roots.
+func (d *Dirs) walk(roots []string) {
+	for _, root := range roots {
+		d.bfsWalkRoot(root)
+	}
+	close(d.scan)
+}
+
+// bfsWalkRoot walks a single directory hierarchy in breadth-first lexical order.
+// Each Gno source directory it finds is delivered on d.scan.
+func (d *Dirs) bfsWalkRoot(root string) {
+	root = filepath.Clean(root)
+
+	// this is the queue of directories to examine in this pass.
+	this := []string{}
+	// next is the queue of directories to examine in the next pass.
+	next := []string{root}
+
+	for len(next) > 0 {
+		this, next = next, this[:0]
+		for _, dir := range this {
+			fd, err := os.Open(dir)
+			if err != nil {
+				log.Print(err)
```

Comment:
While not directly related to this PR, it's crucial that we prioritize ease-of-use for log additions without over-debating the matter. However, it's also important to prevent the potential for an excessive flood of logs due to an overly permissive policy.

To provide some context, we have three primary types of logs:
- `gnoland` blockchain node logs, intended to store comprehensive information for debugging production issues.
- Command-line logs, mainly for warnings.
- Debug logs, which may be verbose and require manual enabling.

Maybe it's time to open a longer RFC issue :)

Reply:
The case where this log is actually called will generally be when we don't have permissions to view a directory. The closest case where I think this can happen is where we're asking `go doc` inside a module (after module support is added), and there are directories inside the module which are actually used ie. for database data or inside of docker, so the fs permissions wouldn't allow you to view them.

From the `gno doc` command perspective, I think this makes sense. If we want to make this cleaner we could instead add function `NewDirsWithErrCallback`, accepting additionally a `func(error)`, so that the log.Print behaviour is specified specifically for the doc command and this package is kept "silent"

Reply:
Yeah, regardless of which way we want to go I think it's out of the scope of this PR, so I'm resolving the convo